enterprise/server/bes_artifact_uploader/bes_artifact_uploader.go

```diff
@@ -0,0 +1,104 @@
+package bes_artifact_uploader
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/build_event_publisher"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"golang.org/x/sync/errgroup"
+
+	bespb "github.com/buildbuddy-io/buildbuddy/proto/build_event_stream"
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+)
+
+type Result struct {
+	Duration time.Duration
+	Digest   *repb.Digest
+	Path     string
+	Name     string
+	Err      error
+}
+
+// Uploader can be used to asynchronously upload artifacts associated with a
+// build event stream.
+type Uploader struct {
+	ctx context.Context
+	eg  *errgroup.Group
+	// ByteStream client used to upload artifacts.
+	bsClient bspb.ByteStreamClient
+	// Publisher used to publish NamedSetOfFiles events when background uploads
+	// have completed.
+	bep *build_event_publisher.Publisher
+	// Bytestream server host name and port.
+	bytestreamHost string
+	// Remote instance name to be used for uploaded artifacts.
+	instanceName string
+	// Channel where upload results are published (non-blocking).
+	resultsChan chan *Result
+}
+
+func New(ctx context.Context, bsClient bspb.ByteStreamClient, bep *build_event_publisher.Publisher, bytestreamHost, instanceName string) *Uploader {
```

Comment:
Rather than injecting bsClient, bytestreamHost (and instanceName?) could you use flags in this file for those things and construct that stuff in New() (and return an err also)

Reply:
Done